internal/instrumentation/instrumentation.go

```diff
@@ -0,0 +1,258 @@
+		ApiTraffic: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "flightctl_api_requests_api_total",
+		}),
+		AgentTraffic: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "flightctl_api_requests_agent_total",
+		}),
+		SuccessLatency: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "flightctl_api_latencies_success_seconds",
+			Buckets: cfg.Prometheus.ApiLatencyBins,
+		}),
+		ErrorLatency: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    "flightctl_api_latencies_error_seconds",
+			Buckets: cfg.Prometheus.ApiLatencyBins,
+		}),
+		SloViolations: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "flightctl_api_errors_slo_total",
+		}),
+		ClientErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "flightctl_api_errors_client_total",
+		}),
+		ServerErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "flightctl_api_errors_server_total",
+		}),
+		CpuUtilization: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "flightctl_api_cpu_load",
+		}),
+		MemoryUtilization: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "flightctl_api_memory_load",
+		}),
+		DiskUtilization: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "flightctl_api_disk_load",
+		}),
+	}
+}
+
+func NewMetricsServer(
+	log logrus.FieldLogger,
+	cfg *config.Config,
+	metrics *ApiMetrics,
+) *MetricsServer {
+	return &MetricsServer{
+		log:      log,
+		cfg:      cfg,
+		metrics:  metrics,
+		registry: prometheus.NewRegistry(),
+	}
+}
+
+// We need to access the HTTP status code in our instrumentation middleware
+// ResponseWriter does not let us do this, so wrap it in an
+// interface that will catch and save the written status code
+type loggingResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func NewLoggingResponseWriter(w http.ResponseWriter) *loggingResponseWriter {
+	return &loggingResponseWriter{
+		w,
+		0,
+	}
+}
+
+func (lw *loggingResponseWriter) WriteHeader(statusCode int) {
+	lw.statusCode = statusCode
+	lw.ResponseWriter.WriteHeader(statusCode)
+}
+
+func (m *MetricsServer) Run(ctx context.Context) error {
+	m.metrics.RegisterWith(m.registry)
+
+	srv := &http.Server{
+		Addr:         m.cfg.Prometheus.Address,
+		Handler:      promhttp.HandlerFor(m.registry, promhttp.HandlerOpts{Registry: m.registry}),
+		ReadTimeout:  5 * time.Second,
+		WriteTimeout: 10 * time.Second,
+	}
+
+	go m.auditCpuWorker()
+	go m.auditMemoryWorker()
+	go m.auditDiskWorker()
+
+	go func() {
+		<-ctx.Done()
+		m.log.Println("Shutdown signal received:", ctx.Err())
+		ctxTimeout, cancel := context.WithTimeout(context.Background(), gracefulShutdownTimeout)
+		defer cancel()
+
+		srv.SetKeepAlivesEnabled(false)
+		_ = srv.Shutdown(ctxTimeout)
+	}()
+
+	if err := srv.ListenAndServe(); err != nil && !errors.Is(err, net.ErrClosed) {
+		return err
+	}
+
+	return nil
+}
+
+func (m *MetricsServer) auditCpuWorker() {
+	var lastIdle uint64 = 0
+	var lastTotal uint64 = 0
+	for {
```

Comment:
could we consider adding a context and maybe a ticker here . We want to ensure that we close down these go routines when the context is cancelled

```
	ticker := time.NewTicker(5 * time.Second) 
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			m.log.Debug("some details about context cancellation.")
			return
		case <-ticker.C:
		  // do stuff every 5s
```